src/shared/libs/iota/quorum.js

```diff
@@ -0,0 +1,415 @@
+import map from 'lodash/map';
+import filter from 'lodash/filter';
+import includes from 'lodash/includes';
+import isEmpty from 'lodash/isEmpty';
+import isUndefined from 'lodash/isUndefined';
+import transform from 'lodash/transform';
+import size from 'lodash/size';
+import sampleSize from 'lodash/sampleSize';
+import union from 'lodash/union';
+import uniq from 'lodash/uniq';
+import IOTA from 'iota.lib.js';
+import { isNodeSynced } from './extendedApi';
+import {
+    QUORUM_THRESHOLD,
+    QUORUM_SIZE,
+    QUORUM_NODES_SYNC_CHECKS_INTERVAL,
+    DEFAULT_BALANCES_THRESHOLD,
+} from '../../config';
+import { EMPTY_HASH_TRYTES } from './utils';
+import { findMostFrequent } from '../utils';
+import Errors from '../errors';
+
+/**
+ * Resolves only if size of synced nodes is greater than or equal to quorum size
+ *
+ * @method rejectIfNotEnoughSyncedNodes
+ *
+ * @param {array} nodes
+ * @param {number} quorumSize
+ * @returns {Promise}
+ */
+const rejectIfNotEnoughSyncedNodes = (nodes, quorumSize = QUORUM_SIZE) => {
+    if (size(nodes) < quorumSize) {
+        return Promise.reject(new Error(Errors.NOT_ENOUGH_SYNCED_NODES));
+    }
+
+    return Promise.resolve();
+};
+
+/**
+ *   Determines quorum result for supported methods.
+ *
+ *   @method determineQuorumResult
+ *   @param {array} validResults
+ *   @param {number} faultyResultCount
+ *
+ *   @returns {function}
+ **/
+const determineQuorumResult = (validResults, quorumSize = QUORUM_SIZE) => {
```

Comment:
That's correct. Invalid results are basically the results from unresponsive nodes. 

If there's a network error, we explicitly mark the result as `undefined` [here](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L273) and later on filter these results.

Reply:
To clarify that this is intentional (seems good to me)
the quorum is calculated *only* within the responsive nodes.
this should be mentioned somewhere - b/c it changes the strategy to attack the quorum.

for example:
```
N = 7
m = 3 (malicious)
scenario A:
all m vote No
all honest vote Yes
Result: 4/7= FailSafe

scenario B:
2 m vote "unresponsive"
1 m votes No
all honest vote Yes
Result: 3/4= Yes
```

so the best stratagy is to participate.
OTOH,
if the malicious nodes manage to kick off honest nodes (the will become uresponsive), they can pass bad elections:
```

scenario C:
3 m vote No
2 honest vote Yes
2 honest vote "unresponsive" due to malice
Result: 3/5= No
(well that's cutting it close, but you get the drift)
```

to give a full picture - anouther option would be to have "unresponsive" vote FailSafe.
but propbably needs some analysis as well.

Reply:
> the quorum is calculated only within the responsive nodes.

The quorum is not calculated within the responsive nodes. Unresponsive nodes would also be considered. Explicitly passing `undefined` as a result for unresponsive nodes is just to easily filter out invalid results. 

Quorum [calculation](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L73) would always be out of `N` (`N === quorumSize`). 

> scenario B:
2 m vote "unresponsive"
1 m votes No
all honest vote Yes
Result: 3/4= Yes

In this case, result would be 3/N. 

> anouther option would be to have "unresponsive" vote FailSafe.

IIUC, then always calculating quorum out of `N` basically does the same?

Reply:
> That's correct. Invalid results are basically the results from unresponsive nodes.
so that is confusing - the param `validResults` sounds like it's only the.. well ... valid results. filtering the unresponsive nodes.

Reply:
oh, so `quorumSize = QUORUM_SIZE` is `N`, regardless of the size of `validResults` - :+1: 
maybe worth another pass to make it a bit clearer :)

Reply:
> oh, so quorumSize = QUORUM_SIZE is N, regardless of the size of validResults

That's correct. I'll make another commit to make it clearer. Sorry about the confusion.

Reply:
[This](https://github.com/iotaledger/trinity-wallet/pull/631/commits/d0a4dd9c0a12e24610e6238b13de71fce19a8543) should improve the readability of the overall implementation.